numpy/core/src/private/npy_longdouble.c

```diff
@@ -0,0 +1,99 @@
+#include <Python.h>
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_math.h"
+
+/* This is a backport of Py_SETREF */
+#define NPY_SETREF(op, op2)                      \
+    do {                                        \
+        PyObject *_py_tmp = (PyObject *)(op);   \
+        (op) = (op2);                           \
+        Py_DECREF(_py_tmp);                     \
+    } while (0)
```

Comment:
This would probably be generally handy throughout the codebase.

A job for another PR: would there be a better place to put it, and do we want to cal lit `Py_SETREF` and `#ifndef Py_SETREF` it?

Reply:
If we decide to use it I agree that doing and `#ifndef` seems reasonable.

For anyone else reading, it looks like this python macro was created in python3.4, and there are a lot of discussions of it on the python mailing list and on the python issue tracker. It is still going through some changes as of april 2016, I think they decided to split it into two macros `Py_SETREF` and `Py_XSETREF`.

One motivation for it is to avoid a certain kind of bug, according to the [comment in CPython](https://github.com/python/cpython/blob/master/Include/object.h#L856). On the other hand it's not part of the "stable api", so there are some risks in using it.